File: src/java/SmithNumber.java
package java;

import java.util.ArrayList;

public class SmithNumber {

    private static int countDigitsSum(long num) {
        int sum = 0;

        while(num > 0) {
            sum += num % 10;
            num /= 10;
        }

        return sum;
    }

    private static ArrayList<Long> returnPrimeFactors(long num) {
        ArrayList<Long> primeFactors = new ArrayList<>();

        for (long i = 2; i * i <= num ; i++) {
            while (num % i == 0) {
                primeFactors.add(i);
                num /= i;
            }
        }
        if(num > 1)
            primeFactors.add(num);

        return primeFactors;
    }

    public static boolean isSmithNumber(long num) {
        int numDigitSum = countDigitsSum(num);
        int factorDigitSum = 0;

        ArrayList<Long> primeFactors = returnPrimeFactors(num);
        if(primeFactors.size() == 1)
            return false;

        for(long val : primeFactors)
            factorDigitSum += countDigitsSum(val);

        return factorDigitSum == numDigitSum;
    }

}

File: src/C++/SmithNumber.cpp
#include<iostream>
#include <vector>
using namespace std;

int getDigitSum(int num){
    int sum = 0;
    int digit = 0;

    while(num > 0){
        digit = num % 10;
        sum += digit;
        num /= 10;
    } 
    
    return sum;
}

vector<int> getPrimeFactors(int num){
    vector<int> primeFactors = {};
    for (int i = 2;i*i <= num;i++){
        while( num % i == 0){
            primeFactors.push_back(i);
            num /= i;
        }
    }
    if( num > 1) {
        primeFactors.push_back(num);
    }

    return primeFactors;
}

bool isSmithNumber(int num){

    int numDigitSum = getDigitSum(num);
    int factorDigitSum = 0;

    vector<int> primes = getPrimeFactors(num);
    if(primes.size() == 1 && primes[0] == num)
        return false;

    for(int i : primes)
        factorDigitSum += getDigitSum(i);

    
    return factorDigitSum == numDigitSum;
}
